Options/Applicative/Common.hs

```diff
@@ -272,10 +274,12 @@ treeMapParser g = simplify . go False False False g
       = Leaf (f (OptHelpInfo m d r) opt)
       | otherwise
       = MultNode []
-    go m d r f (MultP p1 p2) = MultNode [go m d r f p1, go m d r' f p2]
+    go m d r f (MultP p1 p2) =
+      MultNode [go m d r f p1, go m d r' f p2]
       where r' = r || hasArg p1
-    go m d r f (AltP p1 p2) = AltNode [go m d' r f p1, go m d' r f p2]
-      where d' = d || has_default p1 || has_default p2
+    go m d r f p@(AltP p1 p2) =
+      AltNode (has_default p && not (has_default p1 && has_default p2)) [go m d' r f p1, go m d' r f p2]
+      where d' = d || (has_default p && not (has_default p1 && has_default p2))
```

Comment:
This logic is duplicated and getting relatively complex.
Please use a let binding and add an explanatory comment.

Reply:
Done